allennlp/nn/util.py

```diff
@@ -207,17 +220,47 @@ def viterbi_decode(tag_sequence: torch.Tensor, transition_matrix: torch.Tensor):
     viterbi_score : float
         The score of the viterbi path.
     """
-    sequence_length, _ = list(tag_sequence.size())
+    sequence_length, num_tags = list(tag_sequence.size())
+    if tag_evidence:
+        if len(tag_evidence) != sequence_length:
+            raise ConfigurationError("Evidence was provided, but it was not the same length as "
+                                     "the sequence. Found sequence of length: {} and evidence: {}"
+                                     .format(sequence_length, tag_evidence))
+    else:
+        tag_evidence = [-1 for _ in range(sequence_length)]
+
     path_scores = []
     path_indices = []
-    path_scores.append(tag_sequence[0, :])
+
+    if tag_evidence[0] != -1:
+        one_hot = torch.zeros(num_tags)
+        one_hot[tag_evidence[0]] = 100000.
```

Comment:
I think it'd be better to define some constants, `VERY_LARGE`, and `VERY_NEGATIVE`, in `common.util` or `nn.util`, and use those instead of hard-coding these values throughout these functions.  I _thought_ we had already done this, but I can't find those constants at the moment.

Reply:
I was caught in a bit of a dilemma here, because we need to make these observed potential functions to be larger than any transition potential to any other tag, but at the same time, it's not very helpful if you are actually using the path scores for some purpose to have them unnecessarily inflated, for instance if you don't have any transition constraints. I couldn't come up with a good way to get around this, do you have any ideas?

Reply:
The caller has the logits and the transition matrix, so they can recover the original path scores themselves if they really want them.  Or, alternatively, if you use a known constant, they could subtract/add the value of that constant in the appropriate places to recover the scores, assuming the constant wasn't large enough that it washed out the original score.